text/0000-pod.md

```diff
@@ -0,0 +1,195 @@
+setting of a socket:
+
+```rust
+fn getsockopt(sockfd: c_int, optname: c_int, optval: &mut [u8]) -> Result;
+```
+
+The `linger` function cannot be written today because there is no
+`mem::as_mut_bytes` function. Instead, the user would most likely create a slice
+ad hoc with unsafe functions and manual size calculations.
+
+Consider the case of C unions:
+
+Various proposals have been made how to express C unions in rust code. The most
+recent proposal suggested restricting the types in such unions to `Copy` in
+order to avoid various sources of unsafety. However, such unions would still
+have to be `unsafe` since, for example, `&u8` is `Copy`  but cannot contain
+arbitrary data.
+
+Note that C unions only ever contain plain old data (as defined below) since all
+possible C types are plain old data. Hence, if the rust equivalent of C unions
+were restricted to plain old data, using it would be completely safe.†
+
+(† Some questions regarding the existence of `undef` fields in large variants
+remain.)
+
+Lastly, a plain old data type allows for the following simple implementation of
+a random value generator:
+
+```rust
+pub trait RandomValueGenerator {
+    fn generate_random_bytes(&mut self, bytes: &mut [u8]);
+
+    fn generate<T: Pod>(&mut self) -> T {
+        let mut t = mem::zeroed();
+        self.create_random_bytes(mem::as_mut_bytes(&mut t));
+        t
+    }
+}
+```
+
+## Various convenience functions using plain old data
+
+This section contains a list of useful functions that can be written with a
+plain old data trait.
+
+```rust
+/// Creates an object that has all bytes set to zero.
+fn zeroed<T: Pod>() -> T;
+```
+
+```rust
+/// Returns the mutable in-memory representation of an object.
+fn as_mut_bytes<T: Pod>(val: &mut T) -> &mut [u8];
+```
+
+```rust
+/// Turns a slice into a reference to a Pod type if it's suitable.
+///
+/// = Remarks
+///
+/// The buffer is suitable if it is large enough to hold the type and properly
+/// aligned.
+pub fn from_bytes<T: Pod>(buf: &[u8]) -> Option<&T>;
+
+pub fn from_mut_bytes<T: Pod>(buf: &mut [u8]) -> Option<&mut T>;
+```
+
+```rust
+impl<T> [T] {
+    /// Returns a mutable byte slice covering the same range as the slice.
+    pub fn as_mut_bytes(&mut self) -> &mut [u8] where T: Pod;
+}
+```
+
+# Detailed design
+[design]: #detailed-design
+
+Add a `pod` lang item and safe `Pod` trait to mark types which are valid when
+they contain arbitrary bit patterns.
+
+The following types are **Pod candidates**:
+
+* `u8`, `u16`, `u32`, `u64`, `usize`,
+* `i8`, `i16`, `i32`, `i64`, `isize`,
+* `f32`, `f64`,
+* raw pointers,
+* arrays of `Pod` types,
+* tuples of `Pod` types, and
+* structs where all fields are public and `Pod`.
+
+Only types that are Pod candidates can implement the `Pod` trait. Arrays
+of `Pod` types and tuples of `Pod` types automatically implement the `Pod`
+trait.
+
+Structs with `Drop` implementations are not Pod candidates since the drop flag
+is considered a private field.
```

Comment:
Also because `Pod` implies `Copy` (just like `Copy` implies `Clone`), and `Copy` types are never `Drop`.